### PR TITLE
chore: Adds lazy loading and fetchOnlyOnSearch to the Select component

### PR DESCRIPTION
**Pull request by michael-s-molina**
### SUMMARY
- Adds lazy loading and fetchOnlyOnSearch to the Select component.
- Changes the dataset select of the native filters to use lazy loading

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
https://user-images.githubusercontent.com/70410625/126681711-07d2c182-619e-404d-882a-ad1a4efb1a39.mp4

https://user-images.githubusercontent.com/70410625/126326949-e6b55b84-8765-4a8f-952b-4a161e2dea8a.mov

https://user-images.githubusercontent.com/70410625/126326962-1b259658-cbba-42a0-84a5-a69787aa18ae.mov

### TESTING INSTRUCTIONS
Native filters test:
- Open the native filters modal
- Check that the dataset select is working as intended
- Check if the other fields that depend on the dataset select are working as intended

Storybook test:
- Open the Storybook
- Open the AsyncSelect
- Check that queries are only triggered after interacting with the select
- Enable `fetchOnlyOnSearch` control
- Check that queries are only triggered after a search input

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/components/Select/Select.tsx
Comment (geido) [inline superset-frontend/src/components/Select/Select.tsx:146]: What do you think of changing this prop to `fetchOnFocus`? I think it is more accurate as you are not typing anything but only focusing the input.

File: superset-frontend/src/components/Select/Select.tsx
Comment (geido) [inline superset-frontend/src/components/Select/Select.tsx:364]: Are you sure that we should remove this one? As far as I remember, this should check whether the option exists to avoid unnecessary requests against the backend.

File: superset-frontend/src/components/Select/Select.tsx
Comment (geido) [inline superset-frontend/src/components/Select/Select.tsx:380]: Do we really need to check whether `loadingEnabled` is `true` here?

**Review by geido — COMMENTED** [review]
LGMT! Just some minor feedback and a question regarding whether we should keep this option open or just make it the standard.

File: superset-frontend/src/components/Select/Select.tsx
Comment (michael-s-molina) [reply superset-frontend/src/components/Select/Select.tsx:146]: > What do you think of changing this prop to `fetchOnFocus`? I think it is more accurate as you are not typing anything but only focusing the input.
>
> Also, I remember agreeing that we should just set this as the standard as it might be dangerous for performance otheriwse. What do you think?

I think you misunderstood the code. By default, the queries will only be triggered when the select is opened for the first time. This is controlled by the `loadingEnabled` state. 

The `fetchOnlyOnSearch` property is another behavior where the queries will only be triggered when there's a search input present.

File: superset-frontend/src/components/Select/Select.tsx
Comment (michael-s-molina) [reply superset-frontend/src/components/Select/Select.tsx:380]: Yes. Because no query should be fired before the user opens the select for the first time.

File: superset-frontend/src/components/Select/Select.tsx
Comment (michael-s-molina) [reply superset-frontend/src/components/Select/Select.tsx:364]: Yes. The problem with this check is that a search can result in more than one page, which means that we can have results that were not fetched yet and we should fire the request.

File: superset-frontend/src/components/Select/Select.tsx
Comment (geido) [reply superset-frontend/src/components/Select/Select.tsx:146]: You are right. I was under the false impression this would only solve the lazy loading. Thanks for taking care of both cases!

**Review by geido — APPROVED** [review]
Thanks for answering my questions. LG(reat)TM!

**Comment by michael-s-molina** [comment]
Just waiting for QA approval to merge. @jinghua-qa @junlincc @rosemarie-chiu @adam-stasiak

**Review by rosemarie-chiu — APPROVED** [review]
LGTM on UX
CleanShot 2021-07-22 at 11 11 37

**Pull request merged**
